### PR TITLE
Add source query validation for dictionaries

### PR DESCRIPTION
**Pull request by alexkats**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added source query validation when ClickHouse is used as a source for a dictionary.

<details><summary>CI</summary>

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

</summary>

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 9626c38e2965300bd4ba8ec458689e3e1f09b7c8 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72548/9626c38e2965300bd4ba8ec458689e3e1f09b7c8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/72520/b6725d33d115bb3b5992388b6c67600389ae0843/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 1]
```diff
@@ -1437,16 +1437,16 @@ namespace
 void addTableDependencies(const ASTCreateQuery & create, const ASTPtr & query_ptr, const ContextPtr & context)
 {
     QualifiedTableName qualified_name{create.getDatabase(), create.getTable()};
-    auto ref_dependencies = getDependenciesFromCreateQuery(context->getGlobalContext(), qualified_name, query_ptr, context->getCurrentDatabase());
-    auto loading_dependencies = getLoadingDependenciesFromCreateQuery(context->getGlobalContext(), qualified_name, query_ptr);
+    auto ref_dependencies = getDependenciesFromCreateQuery(context->getGlobalContext(), qualified_name, query_ptr, context->getCurrentDatabase(), /*can_throw*/ true);
```
Comment (tavplubix) [inline src/Interpreters/InterpreterCreateQuery.cpp:1440]: Looks like we call `addTableDependencies` after calling doCreateTable (which add the new table/dictionary to the database, and contains something like a "commit point"): https://github.com/ClickHouse/ClickHouse/blob/e7e30d67fb2e2fdd603498f04f2aed5f11d48e0f/src/Interpreters/InterpreterCreateQuery.cpp#L1694-L1701

So it's too late to check if the query is valid or not, the table is already created at this moment

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 2]
```diff
@@ -1437,16 +1437,16 @@ namespace
...
+    auto loading_dependencies = getLoadingDependenciesFromCreateQuery(context->getGlobalContext(), qualified_name, query_ptr, /*can_throw*/ true);
     DatabaseCatalog::instance().addDependencies(qualified_name, ref_dependencies, loading_dependencies);
 }
 
 void checkTableCanBeAddedWithNoCyclicDependencies(const ASTCreateQuery & create, const ASTPtr & query_ptr, const ContextPtr & context)
 {
     QualifiedTableName qualified_name{create.getDatabase(), create.getTable()};
-    auto ref_dependencies = getDependenciesFromCreateQuery(context->getGlobalContext(), qualified_name, query_ptr, context->getCurrentDatabase());
-    auto loading_dependencies = getLoadingDependenciesFromCreateQuery(context->getGlobalContext(), qualified_name, query_ptr);
+    auto ref_dependencies = getDependenciesFromCreateQuery(context->getGlobalContext(), qualified_name, query_ptr, context->getCurrentDatabase(), /*can_throw*/true);
```
Comment (tavplubix) [inline src/Interpreters/InterpreterCreateQuery.cpp:1448]: However, the check in `checkTableCanBeAddedWithNoCyclicDependencies` should be sufficient https://github.com/ClickHouse/ClickHouse/blob/e7e30d67fb2e2fdd603498f04f2aed5f11d48e0f/src/Interpreters/InterpreterCreateQuery.cpp#L1861-L1862

File: src/Interpreters/InterpreterCreateQuery.cpp (hunk 1 above)
Comment (alexkats) [reply src/Interpreters/InterpreterCreateQuery.cpp:1440]: Yes, I know, but I though that it still can be useful to make it unified with other calls here (perhaps some other check will throw some error at that point). Having said that, I don't have a really strong opinion here, so I can remove it from here if you think it makes more sense

Comment (tavplubix) [reply src/Interpreters/InterpreterCreateQuery.cpp:1440]: I don't think we should pass can_throw=true here, it makes more sense in `checkTableCanBeAddedWithNoCyclicDependencies`

**Review by tavplubix — APPROVED** [review]

**Comment by alexkats** [comment]
Integration tests (tsan) [3/6] - https://github.com/ClickHouse/ClickHouse/issues/71049

**Pull request merged**